### ISSUE TITLE
Migrating from Flow 0.12 & Mortar 0.20

### ISSUE THREAD

**Issue by Lakedaemon** [square/flow#215]
I'm trying to port some of my apps from Flow 0.12 & Mortar 0.20 to Flow 1.0 alpha 2 but I'm wondering about a few things : 

1) do flow apps still need mortar ? (is there still some mortar benefit versus ServiceFactory ?)
also I was heavily using mortar viewPresenters, is there a better way to do things now ?

2) what about mutable keys ? I used to preserve some state by making keys mutable in flow 0.12. 
Is there a different way to do that now ? 

Say, you have a searchView, the user types a query and you want to persist that text for the next time the activity resumes after pause/destroy or a reboot of the device. 
What is the best practice ?

3) It looks like Flow 1.0 alpha 2 is able to persist history through a KeyParceler.
What if version 1 of a user app presists some history and, after an upgrade to version 2 of the same app, 
it tries to read this history... whith screens that have changed between versions ?

**Comment by Zhuinden** [comment]
1.) Mortar scopes still allow hierarchical scoping, while ServiceFactories just allow you to bind the MortarScopes in to make them "magically available" for the duration of a given "tree-key hierarchy" via `Context.getSystemService()`. Or at least in theory.

As for ViewPresenters, I think they were always a pain in the ass, and I'm glad that Flow took over state persistence. 

2.) Keys should be immutable. State should be bound to the Presenter (or Custom Viewgroup). The key is that you are in the SearchView, the mutable element is the state of the view itself, and should therefore belong to the view.

3.) KeyParceler transforms keys into `Parcelable`. It's an in-memory thing, it persists to Bundle for config change / process death.

**Comment by Lakedaemon** [comment]
I see. Thanks for the answers. It'll help me migrate things forward. 
Also, I was glad to discover the Flow(less) library... I'll follow that too.

**Comment by Zhuinden** [comment]
@Lakedaemon I really need to make it more customizable, but it's driving one of my apps and it's great! :slightly_smiling_face: 

It took a bit of beating to make it consistently never freeze up, to be honest. But it's stable now

**Comment by rjrjr** [comment]
I can't imagine developing a new app without Flow, Dagger, RxJava and Retrofit. In Kotlin. Everything else is optional, including Mortar.

**Comment by Zhuinden** [comment]
I'm with that to a degree, although I keep it at Flow(less), Dagger,  and Retrofit. 

Well, also Realm if possible, but hey :tongue:

**Comment by Lakedaemon** [comment]
About 2, I think that I may have to rewrite history (as it holds immutable keys) : 

One of my use case is a Search Screen, with incremental search, where users can bookmark and jump to searches.

For example : 
1) user jumps to a saved search for "Apple"
Flow.set(Search("apple"))
primary key is Search, secondary key is Apple
2) On this screen, he changes the search to "Apple juice" 
-> incremental search happens on "Apple j", "Apple ju", "Apple jui", ... "Appel juice"
I obviously don't want to change view each time the request changes 
-> I have to rewrite history with a MultiKey where the secondary key doesn't change the view but updates it
3) user clicks on a menu item to go to settings
Flow.set(Settings())
4) user goes back
if I haven't rewritten history, I will have a search for "Apple" and not for "Apple Juice" as Flow will 
use the Search("Apple") key with the mutated saved state of the view 

Or am I misunderstanding something ?

Basically, what I want to achieve is to build with flow something akin to 
a browser screen able to display different url that can mutate
How would you achieve that with Keys ?

**Comment by Zhuinden** [comment]
> 2) On this screen, he changes the search to "Apple juice" 
> -> incremental search happens on "Apple j", "Apple ju", "Apple jui", ... "Appel juice"
> I obviously don't want to change view each time the request changes 
> -> I have to rewrite history with a MultiKey where the secondary key doesn't change the view but > updates it

Why? Why would you store the "incremental search" current state inside the Key?

This is exactly why the `KeyManager` was detached from the `History` stack. So that your History stores you're on search, but the current state is `Apple juice`.

Throw that stuff in a Bundle. That's where it belongs.

**Comment by Lakedaemon** [comment]
But how do you launch a search screen with a custom state ?
Flow.set(Search()) will give you the search screen, 
but won't populate the searchView with the "Apple" search 

Or is it possible to make Flow move to a view and to somehow provide it with a Bundle that holds the custom "search=Apple" state ?

in the same way, sometimes, you need to launch activities/Fragments with Intents/arguments

**Comment by Zhuinden** [comment]
I'd just store the `apple` as the key's initial state, and store the state `Apple jui` in a bundle using the [ForceBundler](https://github.com/Zhuinden/flowless/blob/master/flowless-library/src/main/java/flowless/ForceBundler.java#L11-L26)

**Comment by Lakedaemon** [comment]
Ah I see now. 
Keys can be linked to states and you manage them through a KeyManager... 
(I missed that while reading the flow documentation)
Many thanks for the pointers. :)

**Comment by Zhuinden** [comment]
@Lakedaemon how's life? I migrated a project from Flow 0.12 to Flowless the other day, and state persistence stability went through the roof, while additional boilerplate classes went down a notch (goodbye, FramePathContainerView! Goodbye, Mortar ViewPresenters! You won't be missed.)

**Comment by Lakedaemon** [comment]
I made a first attempt to port my code from Flow 0.12 & Mortar 0.2 to Flow 1.0. It worked quite nicely (and indeed removed some boilercode) but there were a few features that I didn't know how to handle. So I took some time to think about it.

Now, I think that I got my remaining issues figured out, so I'll soon make a second attempt

Migrating to Flow 1.0 should be a big improvement to my app (and states)

**Comment by Zhuinden** [comment]
@Lakedaemon where did you get stuck?

**Comment by dcow** [comment]
I scratched some stuff down while doing my own migration (retrospectively pretty obvious but I couldn't find it written down anywhere). May it help someone: https://gist.github.com/dcow/c5365aec81ca0dadf012c1ad5a847bdb 

My takeaway: it would be really really really really helpful if this project put together a migration guide and better yet a compatibility module. I suspect a lot of teams have codebases where someone (old dev the company couldn't keep around or contracting firm) copy-pasted in some square sample code (or cloned that android flow+mortar+dagger bootstrap repo) that used flow-path to do a flow+mortar+dagger setup with the simplepathcontainer+view etc. and haven't touched things since (or really even bothered to learn how flow worked in the first place). I know it's not strictly your fault (but let's not forget: posting heavily opinionated pieces about how fragments suck and offering a pre-release software solution from a position of industry authority _did_ happen), but the reality is: it is your problem. Right now migration is a big barrier. IMO this project needs to do everything in its power to make that process painless while at the same time actually educating people on how flow works. Not every team has the time or resources to dig in and perform a blind refactor at the level of intimacy that Flow, mortar, and dagger hook into a project with no guidance.

Flow is amazing. It's the perfect blend between practical Android ui development and formal ui state expression. Even our iOS team is starting to model their UX flow structure as states (enumerated types with associated value) and transitions with hooks for the ViewControllers to do the actual controlling of the views during transitions. I too find it hard to imagine writing modern android without flow, some di framework, and Kotlin. But it's still pretty hard to seriously advocate for such a disruptive framework when the waters are muddy, the software is pre-release, and there's little formal community communication or documentation.

**Comment by Zhuinden** [comment]
@dcow I wrote about Flow 0.12 => Flow 1.0-alpha [**here**](https://github.com/Zhuinden/MortarFlowSetup/wiki/Part-4:-The-Transition-(to-Flow-1.0-alpha)) Although I do admit it's quite hidden as I consider it an "obsolete repository".

**Comment by Lakedaemon** [comment]
Thanks for both of your posts. 

At this time, I paused migration from Flow 0.12 to 1 alpha because it would pretty much take me between 1 week and 1 month to do and I don't want to delay public releases for my app, that brings money in. 
I'll probably go with flow 1.0 for my next project that I write from scratch and, later on, when I fully uinderstand flow 1.0 (and have a lot of working code to copy and paste), I'll eventually migrate the old stuff to flow 1.0....eventually because I learnt the hard way the wisdom of "don't fix what isn't broken....and I have a working solution with flow 0.12 (even if it obviously as state restoring issues).

Also, I'm waiting for a n official 1.0 release. I don't have the time and the energy to make ports every 6 months (also it costs money).

**Comment by Zhuinden** [comment]
@Lakedaemon I don't know if you'll find it helpful, but I posted [here](https://github.com/square/flow/issues/225#issuecomment-277508674) about a new library I've written, heavily based on how Flow 0.9 and how Flow 1.0-alpha works.

The library is called [**simple-stack**](https://github.com/Zhuinden/simple-stack), and it works just like Flow in terms of navigation - all of Flow's tests pass for it.

I'd argue it's more stable than Flow 1.0-alpha, and much simpler. If you have immutable parcelable keys, then the mapping from one library to another is close to "one-to-one", just the names are slightly different.

**Comment by dcow** [comment]
@Zhuinden I did find your stuff during the migration and it was very helpful. Thank you! I think I'm mostly commenting on the fact that I had to search "flow migration" to even find relevant content and yours is really the only decent stuff out there aside from that sample repo by one of the devs.

**Comment by Zhuinden** [comment]
@dcow glad it helped! :)

-----

I think if you aren't directly using Flow's `ServiceFactory`, then migrating over to my new library is a better option.

Well, either `simple-stack` or `flowless`, really. Simple-stack is less quirky (by which I mean the backstack existing only in `onPostCreate()` but being configured in `attachBaseContext()`), and flowless is less buggy than Flow 1.0-alpha (but I removed ServiceFactory because it annoyed the heck out of me).

Here's a [commit where I converted the original `flow-sample` into using `simple-stack`.](https://github.com/Zhuinden/simple-stack/commit/5cb8ad13e0bc621f3e2ec2225c205997da7c6234)

**Comment by Lakedaemon** [comment]
In a first experiment I had a partially working solution, but it was an hybrid Flow 1.0 alpha for keys and screen changes and Mortar for view Presenters. 

In this first experiment, migration wasn't complete because : 
1) at the time I didn't see how to implement some features : mutable state with  flow 1.0alpha (now I know), dual screens (with dual presenters), redirections...
2) it was an hybrid solution that wasn't using the benefits of flow 1.0 alpha (screen state save/restore). So It always built all data from scratch (I had difficulties making observables work by making them static and tying them to the presenter) in the mortar presenters onLoad. So I didn't even have the benefits of mortar (don't rebuild on screen rotation)


What prevents me now of switching to flow 1.0alpha (or another similar solution) is that (instead of rebuilding things from scratch) I'll have to implement countless parcelable-related methods to save and restore that on screen/configuration  my screens with list adapters and lots of various data backed by complex models change.

I should do it, and I somehow know how but the amount of work is putting me off...

Also, there is no non trivial sample for flow (like a screen with a list adapter that displays complex spanned tied to complex objects) that showcase complex state restoration 
(what if the list is built incrementally async, etc... with scroll to see more...)

**Comment by Zhuinden** [comment]
> (what if the list is built incrementally async, etc... with scroll to see more...)

that's a good question, although the answer is that the data models (`DataManager`) should be singleton and independent from whether the views exist or not.

I typically had the luxury of just throwing them into the Realm, and it gave me autoupdates.

**Comment by loganj** [comment]
@zhuinden's first response is correct.

It's worth mentioning here for future readers that this project is still in _alpha_. There's no migration guide because none of this is stable at all. We're definitely not recommending that you adopt it if you're not comfortable with some API thrash, some brokenness, and lack of things like a migration guide. If you are comfortable with that, then feedback is definitely welcome! :)
